src/syntax/syntax.rkt

```diff
@@ -83,32 +112,130 @@
     (error 'operator-info "Unknown operator ~a" op))
   (hash-ref operators-to-impls op))
 
+;; Checks an "accelerator" specification
+(define (check-accelerator-spec! name itypes otype spec)
+  (define (bad! fmt . args)
+    (error name "~a in `~a`" (apply format fmt args) spec))
+
+  (define (type-error! expr actual-ty expect-ty)
+    (bad! "expression `~a` has type `~a`, expected `~a`" expr actual-ty expect-ty))
+
+  (define-values (vars body)
+    (match spec
+      [`(,(or 'lambda 'λ) (,vars ...) ,spec)
+       (for ([var (in-list vars)])
+         (unless (symbol? var)
+           (bad! "expected symbol `~a` in `~a`" var spec)))
+       (values vars spec)]
+      [_ (bad! "malformed specification, expected `(lambda <vars> <expr>)`")]))
+
+  (unless (= (length itypes) (length vars))
+    (bad! "arity mismatch; expected ~a, got ~a" (length itypes) (length vars)))
+
+  (define env (map cons vars itypes))
+  (define actual-ty
+    (let type-of ([expr body])
+      (match expr
+        [(? number?) 'real]
+        [(? symbol?)
+         (cond
+           [(assq expr env)
+            =>
+            cdr]
+           [else (bad! "unbound variable `~a`" expr)])]
+        [`(if ,cond ,ift ,iff)
+         (define cond-ty (type-of cond))
+         (unless (equal? cond-ty 'bool)
+           (type-error! cond cond-ty 'bool))
+         (define ift-ty (type-of ift))
+         (define iff-ty (type-of iff))
+         (unless (equal? ift-ty iff-ty)
+           (type-error! iff iff-ty ift-ty))
+         ift-ty]
+        [`(,op ,args ...)
+         (unless (operator-exists? op)
+           (bad! "expected operator at `~a`, got `~a` in `~a`" expr op))
+         (define itypes (operator-info op 'itype))
+         (for ([arg (in-list args)] [itype (in-list itypes)])
+           (define arg-ty (type-of arg))
+           (unless (equal? itype arg-ty)
+             (type-error! arg arg-ty itype)))
+         (operator-info op 'otype)]
+        [_ (bad! "expected an expression, got `~a`" expr)])))
+
+  (unless (equal? actual-ty otype)
+    (type-error! body actual-ty otype)))
+
+;; Applies a substitution.
+;; Slightly different than `replace-vars` in `programs.rkt`.
+(define (replace-vars expr env)
+  (let loop ([expr expr])
+    (match expr
+      [(? number?) expr]
+      [(? symbol?) (cdr (assq expr env))]
+      [`(,op ,args ...) `(,op ,@(map loop args))])))
+
+;; Expands an "accelerator" specification.
+;; Any nested accelerator is unfolded into its definition.
+(define (expand-accelerator-spec spec)
+  (let loop ([expr spec])
+    (match expr
+      [(? number?) expr]
+      [(? symbol?) expr]
+      [`(,(? operator-accelerator? op) ,args ...)
+       (define spec (operator-info op 'spec))
+       (match-define `(,(or 'lambda 'λ) (,vars ...) ,body) spec)
+       (define env (map cons vars (map loop args)))
+       (replace-vars body env)]
```

Comment:
I notice that an accelerator can’t expand into an accelerator.i think that’s right.